fawltydeps/traverse_project.py

```diff
@@ -51,13 +77,29 @@ def find_sources(  # pylint: disable=too-many-branches
         else:  # must traverse directory
             traversal.add(path, DepsSource)
 
+    for path in settings.pyenvs if PyEnvSource in source_types else []:
+        # exceptions raised by validate_pyenv_source() are propagated here
+        package_dirs: Optional[Set[PyEnvSource]] = validate_pyenv_source(path)
+        if package_dirs is not None:  # Python environment dir given directly
+            logger.debug(f"find_sources() Found {package_dirs}")
+            yield from package_dirs
+            traversal.ignore(path)  # disable traversal of path below
+        else:  # must traverse directory to find Python environments
+            traversal.add(path, PyEnvSource)
+
     for _cur_dir, subdirs, files, extras in traversal.traverse():
         for subdir in subdirs:  # don't recurse into dot dirs
             if subdir.name.startswith("."):
                 traversal.ignore(subdir)
 
         types = {t for t in extras if t in source_types}
         assert len(types) > 0
+        if PyEnvSource in types:
```

Comment:
In one `for` above we ignore "dot" subdir directories.  We would not like to ignore them for virtual environments. Althought, the probability that someone put virtual env in a subdir is rather low.

Reply:
And what if the logic was a bit different.

Let's divide folders to two categories:
- regular
- hidden (ignored with traversal)
We have 3 types of entities we search for:
- `PyEnvSource`
- `CodeSource`
- `DepsSource`

We have a following martix to search:
|   |  `PyEnvSource` |  `CodeSource` |  `DepsSource` |
|---|---|---|---|
| regular |  x | x | x |
| hidden  |  x |   |   |

We could first check regular folders and then dot folders only for `PyEnvSource`. For that `traversal.ignore` should be modified to keep the list of directories ignored due to the fact that they started with a dot.

Additionally, we also know that some of the hidden folders will not contain virtual envs, and 
```
            if subdir.name.startswith("."):
```
could become
```
IGNORED_DIRS = [".git", ".node",  ]
            if any(subdir.name.startswith(ignored) for ignored in IGNORED_DIRS):
```
and we would never need to recurse to those directories. Still we could use mechanism of revisiting ignored dirs for `PyEnvSource`.

Reply:
To be clear, this is not a blocker. I do not think there is a lot of such corner cases. We could move this Idea to an issue and solve it in a separate PR. The current logic is good enough :)

Reply:
Agreed. I posted https://github.com/tweag/FawltyDeps/issues/226#issuecomment-1635812164 for now, to not forget about this when we get around to refactoring our ignore logic.